File: 563-binary-tree-tilt/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTilt(self, root: Optional[TreeNode]) -> int:
        self.res = 0
    
        def post_order_visit(root):
            if root is None: return 0
            left_v = post_order_visit(root.left)
            right_v = post_order_visit(root.right)
            self.res += abs(left_v - right_v)
            return left_v + right_v + root.val

        post_order_visit(root)
        return self.res

File: 563-binary-tree-tilt/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int findTilt(TreeNode* root) {
        postOrderVisit(root);
        return res;
    }

    int postOrderVisit(TreeNode* root) {
        if (root == nullptr) return 0;

        int left = postOrderVisit(root->left);
        int right = postOrderVisit(root->right);
        res += abs(left - right);
        return left + right + root->val;
    }
};